source/common/http/conn_manager_impl.cc

```diff
@@ -471,19 +471,39 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
   // Make sure we are getting a codec version we support.
   Protocol protocol = connection_manager_.codec_->protocol();
   if (protocol == Protocol::Http10) {
-    // The protocol may have shifted in the HTTP/1.0 case so reset it.
-    request_info_.protocol(protocol);
-    HeaderMapImpl headers{
-        {Headers::get().Status, std::to_string(enumToInt(Code::UpgradeRequired))}};
-    encodeHeaders(nullptr, headers, true);
-    return;
+    // Send "Upgrade Required" if HTTP/1.0 support is not expliictly configured on.
+    if (!connection_manager_.config_.http1Settings().accept_http_10_) {
+      // The protocol may have shifted in the HTTP/1.0 case so reset it.
+      request_info_.protocol(protocol);
+      HeaderMapImpl headers{
+          {Headers::get().Status, std::to_string(enumToInt(Code::UpgradeRequired))}};
+      encodeHeaders(nullptr, headers, true);
+      return;
+    } else {
+      // HTTP/1.0 defaults to single-use connections. Make sure the connection
+      // will be closed unless Keep-Alive is present.
+      state_.saw_connection_close_ = true;
```

Comment:
This is a little broken due to how we are determining HTTP/1.0 in the codec right now: https://github.com/envoyproxy/envoy/blob/master/source/common/http/http1/codec_impl.cc#L357. I don't think it honestly matters if we fix this or not, but it's probably at least worth a comment. Up to you if you want to actually fix or not to now detect versions other than 1.0 and 1.1.

Reply:
One of my favorite debugging tools is

telnet www.google.com 80
GET /

which results in an HTTP/1.0 response.  Id' love to have parity with Envoy.    I'll update the envoy doc side to comment that this feature gets us best-effort HTTP 0.9 as well and add a test.

Reply:
FYI as of https://github.com/envoyproxy/envoy/pull/2720, http_parser will let through arbitrary versions (0.3, 1.3, etc.). We should probably comment this fact somewhere now that we are not generically upgrading non 1.1?

Reply:
Oh, hm.  I'm not particularly interested in supporting non-standard versions.  I'm not averse.  Do you think I should block weird ones or just pass through?

Reply:
I would be fine with just treating everything that is not 1.1 as 1.0/0.9, and just comment to that effect. I don't think it's worth the effort to block the other versions, but up to you.

Reply:
Please. please, don't add any support for HTTP/0.9. It adds a ton of edge-cases (like `GET /` being valid request and responses without headers being valid as well) while providing no benefits at all.

Reply:
It benefits me by allowing traffic we see in prod.

I've compromised on keeping HTTP1.1 pure, please accept that some proxies handle old HTTP and that proxies that opt into HTTP other than HTTP/1.1 are more likely to need 0.9

Reply:
I agree with @alyssawilk. As long as the non-standard functionality is opt-in I think it's reasonable to support.